### PR TITLE
Query block: move patterns modal to dropdown on block toolbar

### LINKED ISSUE #63497: Use "Design" panel for Query Loop block
I propose that we update the Query Loop block to the existing "Design" Inspector panel UI/UX, rather than the "Replace" modal that is currently in place. This would bring the experience to match what is currently in place for header and footer template parts, and also reduce confusion for the "Replace" control, which is used on media block toolbars. 

The "Design" panel may need to be closed by default. 

### Visual

CleanShot 2024-07-12 at 14 20 13

### PR DESCRIPTION
**Pull request by ramonjd**
## What?

Resolves https://github.com/WordPress/gutenberg/issues/63497

This PR:
- replaces the "Replace" button in the block toolbar with a "Change design" button. See: https://github.com/WordPress/gutenberg/pull/66993#issuecomment-2495499304
- and then displays patterns selection content in a drop down

## How?

Extracts the belly from the modal to make it reusable.
Also abstracts the `hasPatterns` test for consistency.

## Testing Instructions

1. Fire up the site editor, and open a template with an existing query block, or create one yourself! TT5 theme is a worthy choice.
2. Then select the query block. Check out the block toolbar. Do you see the "Change design" button? Click on it to display the patterns. 
3. Replace a pattern, make sure that works. 
4. Now narrow the viewport width and ensure it's usable in mobile view.
5. Thank you.

## Screenshots or screencast 

https://github.com/user-attachments/assets/dd9b19d5-b56c-4723-a844-194ae98c14eb

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice one! This is working as intended:

2024-11-14 09 35 29 localhost 3a07410ad554

I can't seem to click this "Choose" button, though, it appears to do nothing for me:

Screenshot 2024-11-14 at 09 36 00

Also a _separate note:_ it would be nice if there was a way for this _design_ panel to start out uncollapsed:

Screenshot 2024-11-14 at 09 35 32

It's probably closed because of an issue loading the previews, you think?

**Comment by ramonjd** [comment]
Thanks for the fast review! It's not quite ready yet, just a shell of a PR 😄 

> I can't seem to click this "Choose" button, though, it appears to do nothing for me:

Oh, that should open the modal right? I think I unhooked that completely by accident. I'll have to reinstate.

> Also a separate note: it would be nice if there was a way for this design panel to start out uncollapsed:
> It's probably closed because of an issue loading the previews, you think?

It's possible as far I know. I left it collapsed because the list is very looooong. But I guess we could overflow-y + scroll? What do you think?

Once I've tidied things up I'll add a better test description 😄

**Comment by jasmussen** [comment]
> Oh, that should open the modal right? I think I unhooked that completely by accident. I'll have to reinstate.

It does, yes.

**Comment by jasmussen** [comment]
> It's possible as far I know. I left it collapsed because the list is very looooong. But I guess we could overflow-y + scroll? What do you think?

It's not for this PR; it's a question that's been on my mind also for templates. The contents of that Design panel is _hugely valuable_, so the fact that it's closed reduces prominence that would otherwise be good to have. 

If we were able to open it (again, probably best to keep that separate and do that across _all_ design panels), the sidebar already scrolls, so perhaps the solution to it being long is just to keep the Design panel as the last one (just before Advanced), so it doesn't push down other controls.

**Comment by ramonjd** [comment]
> The contents of that Design panel is hugely valuable, so the fact that it's closed reduces prominence that would otherwise be good to have.

Thanks for the feedback. I've left it open for now. Let me know what you think.

Once thing I changed about the block inspector patterns list was to show tooltips rather than the titles, which normally appear underneath the images. I found the titles personally confusing, and wasn't sure which pattern they referred to in a long vertical list.

Screenshot 2024-11-15 at 12 02 26 pm

Can go either way on it. I'll take your advice on that one.

File: packages/block-library/src/query/edit/inspector-controls/index.js
Comment (ntsekouras) [inline packages/block-library/src/query/edit/inspector-controls/index.js]: The proper check for existing patterns is by using `useBlockNameForPatterns`. If for some reason we don't have `core/query` patterns but we have patterns for a specific Query Loop variation they wouldn't show.

**Comment by ntsekouras** [comment]
Thanks for the PR! 

If I remember correctly we have the `design` panel closed mostly for avoiding to render many patterns (performance) as it might not be the most common thing to do when editing. 

Despite where we end up (closed/opened) it seems to me it fits better below the main block's control.

Finally, this implementation for Query Loop has also a `search` input whereas the other `design` panel don't (template parts etc..). Should we consolidate here and either add or remove the search input. Additionally how useful do you think it is, especially if we don't have visible titles, and if we do they are quite subtle (small font size etc..).

File: packages/block-library/src/query/edit/inspector-controls/index.js
Comment (ramonjd) [reply packages/block-library/src/query/edit/inspector-controls/index.js]: Ah good to know. I just pulled this over from the existing component that I'm replacing:

https://github.com/WordPress/gutenberg/pull/66993/files#diff-168766f853e5687fa469d2eee2d2d9e0c351aed334a937a42f8349a76ab5fe2cL17

**Comment by ramonjd** [comment]
Thank you for helping here!

> If I remember correctly we have the design panel closed mostly for avoiding to render many patterns (performance) as it might not be the most common thing to do when editing.

Good point. I had it closed initially because of this concern.

The patterns load with the site editor so at least there's not an extra API call. 

I'm not sure to be honest. The list of patterns has the potential to be very long. Would it help or annoy users to have it permanently open? 

If I'm tweaking my query block, I would personally find it a bit of a peeve to have to close it to access my query controls. 

> Despite where we end up (closed/opened) it seems to me it fits better below the main block's control.

I had a little look at it appears the [layout inspector control has priority](https://github.com/WordPress/gutenberg/blob/update%2Fquery-block-add-design-to-inspector-tools/packages/block-editor/src/hooks/layout.js#L230-L230), so I guess — I'm not 100% on this yet — we'd need to specify a new group and order it explicitly in the [block inspector component](https://github.com/WordPress/gutenberg/blob/5629744dc7a95ec6675fbdb02dce86c7a4657291/packages/block-editor/src/components/block-inspector/index.js#L100-L100). 

Looking at this, I think there's an argument for having it lower - the query controls are quite useful and the design panel, if left open, has the potential to be very long.

Screenshot 2024-11-19 at 12 58 27 pm

🤷🏻 

> Finally, this implementation for Query Loop has also a search input whereas the other design panel don't (template parts etc..). Should we consolidate here and either add or remove the search input. Additionally how useful do you think it is, especially if we don't have visible titles, and if we do they are quite subtle (small font size etc..).

I left the search control in for no reason in particular. I think it's only useful if you know what you're looking for. The list can be long, so I guess it's useful for theme devs with knowledge of the theme.

For this PR, I could remove it if folks don't feel strongly about it. 

Or, if it goes the other way, it could be added across the other design panels in another PR.

**Comment by ramonjd** [comment]
Putting this in draft for a while while I test some things.

I was chatting with @annezazu about it and it was important to show something in navigation mode (write mode).

See: https://github.com/WordPress/gutenberg/issues/66614

So I'm just hacking about to see if it's possible to create a `navigationMode` slot, but I haven't checked all the scenarios yet. 

Conceptually it works, but it probably breaks a lot of things too. @ntsekouras not a priority, but when you get a chance, what do you think about the requirements? Is there a better way to display controls in the inspector toolbar in write mode? 


https://github.com/user-attachments/assets/65720fa6-ccb3-4512-aed8-2845717d95b9

Thank you!

File: packages/block-editor/src/components/block-inspector/index.js
Comment (ramonjd) [inline packages/block-editor/src/components/block-inspector/index.js]: Are these tests the same? When could they be different?

Here's the selector:

https://github.com/WordPress/gutenberg/blob/a2662dfcc38a3fe6757eb4ec6a307b6f709d13ce/packages/block-editor/src/store/private-selectors.js#L519-L519

File: packages/block-editor/src/components/block-inspector/index.js
Comment (ramonjd) [inline packages/block-editor/src/components/block-inspector/index.js]: Basically what I'm trying to achieve here is a slot for tools that should only appear in write mode.

File: packages/block-library/src/query/edit/query-content.js
Comment (ramonjd) [inline packages/block-library/src/query/edit/query-content.js]: Not sure if this is best way to test for write mode inside a block?

Or should we be rather testing for `isSectionBlock`?

File: packages/block-editor/src/components/block-inspector/index.js
Comment (ntsekouras) [reply packages/block-editor/src/components/block-inspector/index.js]: There is an algorithm to try to determine what a section block is and has to do with zoom-out as well. @getdave might have more info, but I don't think it's a check we should add.

**Comment by ntsekouras** [comment]
Thanks for exploring this!

I don't know in depth all the nuances with all the different modes (editorMode, blockEditingMode, etc..) so I'm not sure I can give good feedback. I'll cc @WordPress/gutenberg-core .

My thoughts would be that for this [Query Loop issue](https://github.com/WordPress/gutenberg/issues/66614), I imagine changes in how we handle blocks with inner blocks in `contentOnly` mode and not regarding the `editorMode` (as here). In combination we would need some better handling for blocks like Site Title etc.. when in `contentOnly`. 

Having a separate slot could work I guess and we would also need slot for block Toolbar, but right now we have some checks in `image` block to render conditionally such controls - though I'm not sure it's the right approach. There is no much difference here since inside Query Loop you're doing a[ conditional rendering](https://github.com/WordPress/gutenberg/pull/66993/files#diff-17253f75062593a5e244507f79cfaa0503bd2d232f64cc66298fbc9346536278R159) - it would just render different controls and not inside the new slot. So, do we need a new slot? 🤔

File: packages/block-editor/src/components/block-inspector/index.js
Comment (draganescu) [reply packages/block-editor/src/components/block-inspector/index.js]: Maybe we should not do it in write mode and instead do it for the block in design mode 1st. I think content only representation of blocks that do not expose role content attributes is a larger thing to consider. There is at least https://github.com/WordPress/gutenberg/issues/65699 to consider as well

File: packages/block-editor/src/components/block-inspector/index.js
Comment (draganescu) [reply packages/block-editor/src/components/block-inspector/index.js]: @youknowriad @ramonjd here is basically trying the original idea I had too for #66753 have a slot where blocks can put things in certain edit modes. The query block puts a design picker, the navigation block shows the list view. But I believe you had some objections?

It seems to be a problem that it does add yet another slot and we're unclear what "content only representation" even means and we rush into something public.

File: packages/block-library/src/query/edit/query-content.js
Comment (draganescu) [reply packages/block-library/src/query/edit/query-content.js]: It's important to figure out if blocks should check for that. It would be nice if the block editor should know what to do based on something about these blocks (what does navigation and query block have in common...).

Write mode is a top level content only lock, and blocks do have content role attributes. They should not have to implement special UI to "support" features of the block editor, if they support it the UI works.

File: packages/block-library/src/query/editor.scss
Comment (draganescu) [inline packages/block-library/src/query/editor.scss:50]: Is this an override of sorts?

File: packages/block-library/src/query/edit/query-content.js
Comment (draganescu) [inline packages/block-library/src/query/edit/query-content.js]: Maybe we shouldn't need this bifurcation and the block editor should know how to handle write mode? The more I say it the less I think it's doable. Even here, query block shows a design picker, navigation block shows a tree of inner blocks.

**Review by draganescu — COMMENTED** [review]
@WordPress/gutenberg-design should we have a unified approach to write mode for blocks that "should work" in write mode but have no content exposed, or is a block by block approach needed?

@ramonjd could we make the PR address design mode 1st?

**Comment by jarekmorawski** [comment]
Would it make sense to make `Design` a separate tab in the Inspector? That'd solve the scrolling issues you identified earlier. It'd also make the designs easy to find and consistent across different blocks, patterns, and template parts.

image

**Comment by jasmussen** [comment]
Funny you should mention it, I had a similar thought this morning:

Screenshot 2024-11-21 at 09 17 33

Intoducing a new tab is something I'd like to move a bit carefully with, in a Plan B kind of sense, because doing so introduces a few variables in terms of discoverability of the other settings. It may work! But it would be good to improve some of the other basics first. E.g. performance if there's anything to get, opening by default, a consistent position (always top, or always bottom), starting out open if possible, addressing the jumpiness on loading, perhaps with animation or skeleton indicators, or even fixed aspect ratios (3:4 for pages, 4:3 for patterns perhaps). Those improvements might still benefit anything in a tab.

**Comment by youknowriad** [comment]
I'm a bit confused :) 

Why are we moving this to the inspector when we recently updated something similar for shuffling in "zoom-out". Should we consolidate what we do here? Or do we want these things to be in "toolbars" in "zoom-out" and in the inspector in the other modes? 

How is this related to "write mode"?

**Comment by annezazu** [comment]
@youknowriad this is how it's related to Write mode: https://github.com/WordPress/gutenberg/issues/66614 We need a better experience when Write mode is turned on and having available patterns to swap between would greatly improve the experience.

**Comment by youknowriad** [comment]
@annezazu Understood, but can we settle on a single pattern for all these similar interactions: Either the "change design" in the toolbar like we did recently in "zoom-out" or the "design" panel in the inspector. Are there any incentives to use different UX.

**Comment by ramonjd** [comment]
> could we make the PR address design mode 1st?

Agree. Thanks for testing @draganescu 

For this PR I'll revert the demo hack, and just have an MVP to resolve https://github.com/WordPress/gutenberg/issues/63497 so that a unified approach to tackle https://github.com/WordPress/gutenberg/issues/66614 has a dedicated space. 

Thanks everyone for the help and discussion.

File: packages/block-library/src/query/editor.scss
Comment (ramonjd) [reply packages/block-library/src/query/editor.scss:50]: It's to compensate for the removal of the search component in the patterns view.

**Comment by annezazu** [comment]
>but can we settle on a single pattern for all these similar interactions: Either the "change design" in the toolbar like we did recently in "zoom-out" or the "design" panel in the inspector. Are there any incentives to use different UX.

I'd really like us to do so too. IMO I lean towards implementing what we did with zoom out here if we can as adding the "design" panel to the sidebar for the query loop block comes with many downsides:

- Buries important settings for the query loop block. This stands in contrast to a template part that doesn't have any settings to change or adjust.
- Shows a smaller preview that makes it harder to select between.
- Only is visible if someone opens the block settings (some don't for a long time! I've seen this in user testing.). 
- Will show all patterns regardless of how many a theme has which can quickly make for an unweildly experience. 

I defer to design here but this is what's on my mind.

**Comment by ramonjd** [comment]
>  I lean towards implementing what we did with zoom out here if we can as adding the "design" panel to the sidebar for the query loop block comes with many downsides:

Thanks @annezazu and to everyone for their feedback.

You make a pretty good case in relation to the downsides. 

I don't think the block inspector sidebar has the real estate. I've a mind to close this PR.

Still, I've updated this PR to reflect the [original issue's](Resolves https://github.com/WordPress/gutenberg/issues/63497) intentions for the sake of testing. 

Happy to close this one until a consistent write-only state for block tools surfaces.

>  Either the "change design" in the toolbar like we did recently in "zoom-out" or the "design" panel in the inspector. Are there any incentives to use different UX.

Another question could be: are there any blocks that require tools in write mode that are currently blocked by the `isSectionBlock` gating condition in https://github.com/WordPress/gutenberg/blob/436c9b506aaedffd3e7438d7251ac7dc43d506eb/packages/block-editor/src/components/block-inspector/index.js#L279-L279?

If there's a comfortable home for designs in the toolbar, and no further requirements, then maybe this PR can be abandoned.

**Comment by talldan** [comment]
> Another question could be: are there any blocks that require tools in write mode that are current blocked by the isSectionBlock gating condition

I think block style variations are shown in the inspector for sections when zoom out is active, possibly also in write mode.

Some blocks also have special toolbar controls, but I think that's just handled within the block via some conditional rendering (image block with its alt text is an example).

**Comment by aaronrobertshaw** [comment]
> I think block style variations are shown in the inspector for sections when zoom out is active

There's some evolving work happening over on https://github.com/WordPress/gutenberg/pull/67140 to add block style variation controls to the block toolbar in zoom out mode.

**Comment by youknowriad** [comment]
> Another question could be: are there any blocks that require tools in write mode that are current blocked by the isSectionBlock gating condition

I think that condition is not "final", in the sense that it's a temporary solution to the fact that we want to hide most block tools in these modes (content-only, write mode, zoom-out) and instead we only show the tools that the framework (block-editor) can provide and not the ones that the block provides. I agree that it has proved to be a challenge. What's clear (from multiple PRs) is that we need to come up with a good system to decide:

 - What tools make sense for sections (content-only, write mode, zoom-out) from a design perspective? So far the answer was very ad-hoc, we want this tool, but not this tool, that one but not the other... The import bit is the **Why that control over that one.** A question for our designers. @richtabor @jasmussen 
 - With the above answer, we can come up with the right API/technical solution to either implement all these tools in the framework, offer a dedicated slot or a flag (prop) to `BlockControls` and `InspectorControls`...

**Comment by annezazu** [comment]
Quickly tested this and noticed that I can't find the write/design tools in the PR now. I can see a collapsible design section under the Query Loop though but worry this might have messed with something else. 

https://github.com/user-attachments/assets/740b612b-58d6-4c3d-a461-afb5ee309462

**Comment by richtabor** [comment]
> What tools make sense for sections (content-only, write mode, zoom-out) from a design perspective? So far the answer was very ad-hoc, we want this tool, but not this tool, that one but not the other... The import bit is the Why that control over that one. A question for our designers. @richtabor @jasmussen

I think both affordances are appropriate, but perhaps we clean up to where we are already headed today (the "Change design" toolbar control)—before adding something new. 

CleanShot 2024-11-23 at 09 32 56

For blocks with alternate patterns defined via block types, like Query, Header, and Footer, I think it's fine to have the "Change design" control available anytime. It's unlikely to be available in zoom out, unless it's also a top-level block. 
The current "Change design" that is top-level on any pattern with category meta should remain only available in zoom out, as it's potentially much more destructive.

**Comment by ramonjd** [comment]
> Quickly tested this and noticed that I can't find the write/design tools in the PR now. I can see a collapsible design section under the Query Loop though but worry this might have messed with something else.

@annezazu The write mode was placed behind an experimental flag in https://github.com/WordPress/gutenberg/pull/67008

Screenshot 2024-11-24 at 6 41 47 pm

Could that be it?

**Comment by fabiankaegy** [comment]
I like the veriation @richtabor showcases in the toolbar much better than the version in the sidebar. The sidebar just takes up too much vertical space and only having a single pattern up at a time makes it very hard fo actually find / compare anything.

**Comment by annezazu** [comment]
>The write mode was placed behind an experimental flag in https://github.com/WordPress/gutenberg/pull/67008

Yes! I am a fool and didn't connect that that happened just before this :D

**Comment by ramonjd** [comment]
> we clean up to where we are already headed today (the "Change design" toolbar control)

I've done this for the query block. 👍🏻 It's a first pass so if you can see opportunities for refinements, let me know. Thank you!

File: packages/block-library/src/query/edit/inspector-controls/index.js
Comment (aaronrobertshaw) [inline packages/block-library/src/query/edit/inspector-controls/index.js]: This link and the `BlockInfo` slot was just removed in https://github.com/WordPress/gutenberg/pull/67189. I take it this was just a leftover from the last rebase?

Comment (ramonjd) [reply packages/block-library/src/query/edit/inspector-controls/index.js]: yeah probably 😄 I probably didn't fetch before. Thanks for spotting

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for all the iterations here @ramonjd 👍 

This is testing well for me. 

✅ Patterns are rendered appropriately via "Change design" in the block toolbar
✅ Replacing patterns works as expected
✅ Change design popover renders and works in mobile view


https://github.com/user-attachments/assets/e9fa7752-5811-45cf-a31f-dc60bac31e74



My understanding is you've addressed all the feedback for this first pass, so to that end I'm happy to approve. Feel free to wait for a second opinion or design approval if you'd like though.

**Comment by ramonjd** [comment]
Thanks a lot for testing @aaronrobertshaw!! 🙇🏻 

I'll leave this one to simmer overnight to see if design folks suggest any last-minute tweaks.

File: packages/block-library/src/query/edit/query-toolbar.js
Comment (youknowriad) [inline packages/block-library/src/query/edit/query-toolbar.js:24]: Not for this PR but how much of this toolbar control can be made generic and absorbed automatically for all blocks in "block-editor"? (just asking whether that's even possible)

Comment (ramonjd) [reply packages/block-library/src/query/edit/query-toolbar.js:24]: I had a poke around and I think it could be genericized in the block-editor package. 

For example the [align hook displays a block control](https://github.com/WordPress/gutenberg/blob/5797f50e839142856763b2152a11a655c70f4c6c/packages/block-editor/src/hooks/align.js#L144-L144).

As far as I can tell the `hasPatterns` check the query block uses relies only on the blocks and block-editor stores. So no core data.

**Pull request merged**
